cmd/raidz_test/raidz_test.c

```diff
@@ -0,0 +1,800 @@
+			o->rto_dsize = 1ULL <<  MIN(24, MAX(9, value));
+			break;
+		case 't':
+			o->rto_sweep_timeout = value;
+			break;
+		case 'v':
+			o->rto_v++;
+			break;
+		case 'S':
+			o->rto_sweep++;
+			break;
+		case 'B':
+			o->rto_benchmark++;
+			break;
+		case 'D':
+			o->rto_gdb = 1;
+			break;
+		case 'T':
+			o->rto_sanity = 1;
+			break;
+		case 'h':
+			usage(B_TRUE);
+			break;
+		case '?':
+		default:
+			usage(B_FALSE);
+			break;
+		}
+	}
+}
+
+#define	DATA_COL(rm, i) ((rm)->rm_col[raidz_parity(rm) + (i)].rc_data)
+#define	DATA_COL_SIZE(rm, i) ((rm)->rm_col[raidz_parity(rm) + (i)].rc_size)
+
+#define	CODE_COL(rm, i) ((rm)->rm_col[(i)].rc_data)
+#define	CODE_COL_SIZE(rm, i) ((rm)->rm_col[(i)].rc_size)
+
+static int
+cmp_code(raidz_test_opts_t *opts, const raidz_map_t *rm, const int parity)
+{
+	int i, eret = 0;
+
+	VERIFY(parity >= 1 && parity <= 3);
+
+	for (i = 0; i < parity; i++) {
+		if (0 != memcmp(CODE_COL(rm, i), CODE_COL(opts->rm_golden, i),
+			CODE_COL_SIZE(rm, i))) {
+			eret++;
+			if (opts->rto_v > D_INFO) {
+				PRINT("\nparity block [%d] different!\n", i);
+			}
+		}
+	}
+	return (eret);
+}
+
+static int
+cmp_data(raidz_test_opts_t *opts, raidz_map_t *rm)
+{
+	int i, eret = 0;
+	int dcols = opts->rm_golden->rm_cols - raidz_parity(opts->rm_golden);
+
+
+	for (i = 0; i < dcols; i++) {
+		if (0 != memcmp(DATA_COL(opts->rm_golden, i), DATA_COL(rm, i),
+			DATA_COL_SIZE(opts->rm_golden, i))) {
+			eret++;
+			if (opts->rto_v > D_INFO) {
+				PRINT("data block [%d] different!\n", i);
+			}
+		}
+	}
+	return (eret);
+}
+
+static void
+corrupt_colums(raidz_map_t *rm, const int *tgts, const int cnt)
+{
+	int i;
+	int *dst;
+	raidz_col_t *col;
+
+	for (i = 0; i < cnt; i++) {
+		col = &rm->rm_col[tgts[i]];
+		dst = col->rc_data;
+		for (i = 0; i < col->rc_size / sizeof (int); i++)
+			dst[i] = rand();
+	}
+}
+
+void
+init_zio_data(zio_t *zio)
+{
+	int i;
+	int *dst = (int *) zio->io_data;
+
+	for (i = 0; i < zio->io_size / sizeof (int); i++) {
+		dst[i] = rand_data[i];
+	}
+}
+
+static void
+fini_raidz_map(zio_t **zio, raidz_map_t **rm)
+{
+	vdev_raidz_map_free(*rm);
+	raidz_free((*zio)->io_data, (*zio)->io_size);
+	umem_free(*zio, sizeof (zio_t));
+
+	*zio = NULL;
+	*rm = NULL;
+}
+
+static int
+init_raidz_golden_map(raidz_test_opts_t *opts, const int parity)
+{
+	int err = 0;
+	zio_t *zio_test;
+	raidz_map_t *rm_test;
+	const size_t total_ncols = opts->rto_dcols + parity;
+
+	if (opts->rm_golden) {
+		fini_raidz_map(&opts->zio_golden, &opts->rm_golden);
+	}
+
+	opts->zio_golden = umem_zalloc(sizeof (zio_t), UMEM_NOFAIL);
+	zio_test = umem_zalloc(sizeof (zio_t), UMEM_NOFAIL);
+
+	opts->zio_golden->io_offset = zio_test->io_offset = opts->rto_offset;
+	opts->zio_golden->io_size = zio_test->io_size = opts->rto_dsize;
+
+	opts->zio_golden->io_data = raidz_alloc(opts->rto_dsize);
+	zio_test->io_data = raidz_alloc(opts->rto_dsize);
+
+	init_zio_data(opts->zio_golden);
+	init_zio_data(zio_test);
+
+	VERIFY0(vdev_raidz_impl_set("original"));
```

Comment:
Just FYI since this is always built with -DDEBUG you could use ASSERT instead since they'll always be promoted to VERIFY.  No need to change anything but something I wanted to mention.